module/zfs/zio_checksum.c

```diff
@@ -179,10 +283,12 @@ zio_checksum_compute(zio_t *zio, enum zio_checksum checksum,
 		else
 			bp->blk_cksum = eck->zec_cksum;
 		eck->zec_magic = ZEC_MAGIC;
-		ci->ci_func[0](data, size, &cksum);
+		ci->ci_func[0](data, size, spa->spa_cksum_tmpls[checksum],
```

Comment:
I'm not sure this is correct. Your template initialization functions seem to initialize a context, not a context template. If my understanding is correct, the context should be modified by the checksum functions, leaving it invalid for future uses. I'm not seeing where the template gets used to make a new context anywhere....